wallet/wallet.go

```diff
@@ -697,6 +697,147 @@ func (w *Wallet) syncWithChain(birthdayStamp *waddrmgr.BlockStamp) error {
+	var birthdayStamp *waddrmgr.BlockStamp
+	birthday := w.Manager.Birthday()
+
+	// We'll start by getting our chain client and database transaction
+	// ready.
+	chainClient, err := w.requireChainClient()
+	if err != nil {
+		return nil, err
+	}
+	tx, err := w.db.BeginReadWriteTx()
+	if err != nil {
+		return nil, err
+	}
+	ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
+
+	// Determine the latest height known to the chain backend. This will
+	// serve as our finishing point when attempting to find our birthday
+	// stamp.
+	var latestCheckptHeight int32
+	if len(w.chainParams.Checkpoints) > 0 {
+		latestCheckptHeight = w.chainParams.
+			Checkpoints[len(w.chainParams.Checkpoints)-1].Height
+	}
+	_, bestHeight, err := chainClient.GetBestBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	logHeight := latestCheckptHeight
+	if bestHeight > logHeight {
+		logHeight = bestHeight
+	}
+
+	log.Infof("Catching up block hashes to height %d, this will take a "+
+		"while...", logHeight)
+
+	// isCurrent is a helper function that we'll use to determine if the
+	// chain backend is currently synced.
+	isCurrent := func(bestHeight int32) bool {
+		switch c := chainClient.(type) {
+		case *chain.NeutrinoClient:
+			return c.CS.IsCurrent()
+		}
+		return bestHeight >= latestCheckptHeight
+	}
+
+	for height := w.Manager.SyncedTo().Height; birthdayStamp == nil; height++ {
+		// If we've reached our best height and we're not current, we'll
+		// wait for blocks at tip to ensure we find a valid birthday
+		// stamp.
+		for height == bestHeight && !isCurrent(bestHeight) {
+			time.Sleep(100 * time.Millisecond)
+			_, bestHeight, err = chainClient.GetBestBlock()
+			if err != nil {
+				tx.Rollback()
+				return nil, err
+			}
+		}
+
+		// Retrieve the current block's timestamp to determine whether
+		// it could be a valid birthday stamp.
+		hash, err := chainClient.GetBlockHash(int64(height))
+		if err != nil {
+			tx.Rollback()
+			return nil, err
+		}
+		header, err := chainClient.GetBlockHeader(hash)
+		if err != nil {
+			tx.Rollback()
+			return nil, err
+		}
+
+		// If it is, we'll set it and persist it to disk, which should
+		// then allow us to break out of the loop.
+		if header.Timestamp.After(birthday) {
+			log.Debugf("Found birthday block: height=%d, hash=%v",
+				height, hash)
+
+			birthdayStamp = &waddrmgr.BlockStamp{
+				Hash:      *hash,
+				Height:    height,
+				Timestamp: header.Timestamp,
+			}
+
+			err := w.Manager.SetBirthdayBlock(
+				ns, *birthdayStamp, true,
+			)
+			if err != nil {
+				tx.Rollback()
+				return nil, err
+			}
+		}
+
+		err = w.Manager.SetSyncedTo(ns, &waddrmgr.BlockStamp{
+			Hash:      *hash,
+			Height:    height,
+			Timestamp: header.Timestamp,
+		})
+		if err != nil {
+			tx.Rollback()
+			return nil, err
+		}
+
+		// Checkpoint our state every 10K blocks.
+		if height%10000 == 0 {
+			if err := tx.Commit(); err != nil {
+				tx.Rollback()
+				return nil, err
+			}
+
+			tx, err = w.db.BeginReadWriteTx()
```

Comment:
Usually scope issues like this could be solved by not capturing the `tx` variable when it is first created:
```go
tx, err := w.db.BeginReadWriteTx()
if err != nil {
	return nil, err
}
defer func() {
        tx.Rollback()
}()
```
but I think maybe that's even more confusing in this case, and we should keep it as is.

Reply:
We still can't defer the `Rollback` since we're within the function closure, which is called on and terminates after every block.